### PR TITLE
Task 1 ideal weight

### PR DESCRIPTION
**Pull request by gbGiorgi**

### CODE REVIEW

File: Task1/ideal_weight.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+class Ideal
+  NUMBER = 110
+  def initialize
+    puts "Enter your name: "
+    @name = gets.chomp.capitalize
+    puts "Enter your height: "
+    @height = gets.chomp.to_f
+  end
+
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:11]: let's add a call method here and run all other methods inside it

Please read how to use Service Objects in Ruby and try to use them in future tasks. We also use them in rails apps

https://blog.appsignal.com/2020/06/17/using-service-objects-in-ruby-on-rails.html
https://www.honeybadger.io/blog/refactor-ruby-rails-service-object/
https://www.toptal.com/ruby-on-rails/rails-service-objects-tutorial

Service objects are single business actions. You can change the name of your public method if you like. I prefer using call

here is an example
https://paste.ofcode.org/9uJVSsEmVrJZNTTsXs6aLQ

File: Task1/ideal_weight.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  def error_message
+    if @height.zero?
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:13]: you can also print this error if the user passes a negative height

File: Task1/ideal_weight.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      warn "incorrect input"
+      exit 1
+    end
+  end
+
+  def weight
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:19]: correct all names in accordance with these recommendations
https://medium.com/selleo/a-subjective-guide-to-naming-stuff-in-ruby-on-rails-classes-b44928b6c49a

**Review by cyberik-dev — CHANGES_REQUESTED** [review]

File: Task1/ideal_weight.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  def initialize
+    puts "Enter your name: "
+    @name = gets.chomp.capitalize
+    puts "Enter your height: "
+    @height = gets.chomp.to_f
+  end
+
+  def error_message
+    if @height.zero?
+      warn "incorrect input"
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:14]: also, you can raise an exception here

raise AttributeError, 'Parameters must be positive numbers!'
https://rollbar.com/guides/ruby/how-to-raise-exceptions-in-ruby/

and handle it in the call method
```ruby
  def call
    validate_parameters
    area
  rescue AttributeError => e
    warn e
  end

File: Task1/ideal_weight.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#ideal weight
+
+class Ideal
```
[the remark this answers is not in the document]
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: i have Replaced the class name according to the file.

File: Task1/ideal_weight.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  NUMBER = 110
```
[the remark this answers is not in the document]
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: I changed both the name of the constant and the name of the class.

File: Task1/ideal_weight.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#ideal weight
```
[the remark this answers is not in the document]
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: I corrected the code according to rubocope notes

File: Task1/ideal_weight.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+class Ideal
+  NUMBER = 110
+  def initialize
+    puts "Enter your name: "
+    @name = gets.chomp.capitalize
+    puts "Enter your height: "
+    @height = gets.chomp.to_f
+  end
+
+  def error_message
```
[the remark this answers is not in the document]
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: corrected all names.

File: Task1/ideal_weight.rb (hunk 1 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb:13]: added call method and changed other methods to private

File: Task1/ideal_weight.rb (hunk 2 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: i have modified method for this error

File: Task1/ideal_weight.rb (hunk 3 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: corrected all names...

File: Task1/ideal_weight.rb (hunk 4 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: yes, however in this moment i have fixed all problems and mistakes in my code, tested all possible inputs and It works much better now than ever.

File: Task1/ideal_weight.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    difference = @height.to_i - SUBSTRACTOR
+    if difference.negative?
+      "#{@name}, your weight is already optimal"
+    else
+      "#{@name}, your ideal weight is #{difference}"
+    end
+  end
+end
+
+puts IdealWeight.new.call
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:39]: good job but correct rubocop's last remark please

Layout/TrailingEmptyLines: Final newline missing.
puts IdealWeight.new.call

File: Task1/ideal_weight.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  def call
+    validate_parameters
+    detect_ideal_weight
+  end
+
+  private
+
+  def validate_parameters
+    return unless @height.to_i <= EDGE || @height.count('a-zA-Z').positive?
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:23]: move 'a-zA-Z' to a constant

File: Task1/ideal_weight.rb [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    warn 'incorrect input for height!'
+    exit 1
+  end
+
+  def detect_ideal_weight
+    difference = @height.to_i - SUBSTRACTOR
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:30]: you can extract this to a method

please read about Ruby Memoization
https://www.honeybadger.io/blog/ruby-rails-memoization/
https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/
https://thoughtbot.com/blog/ruby-memoization-and-alternatives

and use it in this method

File: Task1/ideal_weight.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    if difference.negative?
```
Comment (cyberik-dev) [inline Task1/ideal_weight.rb:31]: You also can use a predicate method here

`if ideal_weight?`

**Review by cyberik-dev — CHANGES_REQUESTED** [review]

File: Task1/ideal_weight.rb (hunk 9 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb:45]: i did it just now

File: Task1/ideal_weight.rb (hunk 10 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: i have created new constant for it

File: Task1/ideal_weight.rb (hunk 11 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: yes i did it just now

File: Task1/ideal_weight.rb (hunk 12 above)
Comment (gbGiorgi) [reply Task1/ideal_weight.rb]: ok

**Pull request merged**